### PR TITLE
[CC-4299] Cluster UI - Transactions Page Connected

### PR DESCRIPTION
**Pull request by nathanstilwell**
WIP

Adding a connected Transactions page for Cloud Console

#### Checklist
- [ ] I have written or updated test for the changes I made
- [ ] I have updated the README of the package I'm working in to reflect my changes
- [ ] I have added or updated Storybook if appropriate for my changes

---
This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/ui/325)

### CODE REVIEW

File: packages/cluster-ui/src/transactionsPage/transactionsPageConnected.tsx [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+import { nodeRegionsByIDSelector } from "../store/nodes";
+
+export const TransactionsPageConnected = withRouter(
+  connect<
+    TransactionsPageProps,
+    TransactionsPageDispatchProps,
+    RouteComponentProps
+  >(
+    (state: AppState) => ({
+      data: selectTransactionsData(state),
```
Comment (nathanstilwell) [inline packages/cluster-ui/src/transactionsPage/transactionsPageConnected.tsx:26]: @elkmaster @koorosh I'm getting a TypeScript error here, 

```
(alias) selectTransactionsData(state: AppState): any[] | cockroach.server.serverpb.StatementsResponse
import selectTransactionsData
No overload matches this call.
  The last overload gave the following error.
    Argument of type '(state: AppState) => { data: any[] | cockroach.server.serverpb.StatementsResponse; nodeRegions: { [key: string]: string; }; error: Error; }' is not assignable to parameter of type 'MapStateToPropsParam<TransactionsPageProps, RouteComponentProps<{}, StaticContext, unknown>, DefaultRootState>'.
      Type '(state: AppState) => { data: any[] | cockroach.server.serverpb.StatementsResponse; nodeRegions: { [key: string]: string; }; error: Error; }' is not assignable to type 'MapStateToPropsFactory<TransactionsPageProps, RouteComponentProps<{}, StaticContext, unknown>, DefaultRootState>'.
        Type '{ data: any[] | cockroach.server.serverpb.StatementsResponse; nodeRegions: { [key: string]: string; }; error: Error; }' is not assignable to type 'MapStateToProps<TransactionsPageProps, RouteComponentProps<{}, StaticContext, unknown>, DefaultRootState>'.
          Type '{ data: any[] | StatementsResponse; nodeRegions: { [key: string]: string; }; error: Error; }' provides no match for the signature '(state: DefaultRootState, ownProps: RouteComponentProps<{}, StaticContext, unknown>): TransactionsPageProps'.ts(2769)
index.d.ts(213, 5): The last overload is declared here.
```

Not convinced I have this right yet, but I don't see anything missing from the Transactions Page props.

**Review by koorosh — COMMENTED** [review]
> Reviewed 8 of 8 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/ui/325)* status: all files reviewed, 3 unresolved discussions (waiting on @elkmaster and @nathanstilwell)

*[a discussion](https://reviewable.io/reviews/cockroachdb/ui/325#-MahWQA-6M_X2XzJlZhb:-MahWQA-6M_X2XzJlZhc:bre50jr) (no related file):*
@nathanstilwell , in general these changes look good to me. 
I added some suggestions to resolve type errors.

---
___
*[packages/cluster-ui/src/store/transactions/transactions.sagas.ts, line 8 at r1](https://reviewable.io/reviews/cockroachdb/ui/325#-MahVqMD3XvoX2RBs2z1:-MahVqMD3XvoX2RBs2z2:b-7yu1mv) ([raw file](https://github.com/cockroachdb/ui/blob/064944432cb7111aebb80304e809173bfd8c0428/packages/cluster-ui/src/store/transactions/transactions.sagas.ts#L8)):*
> ```XML
> import { CACHE_INVALIDATION_PERIOD, throttleWithReset } from "src/store/utils";
> 
> export function* refreshStatementsSaga() {
> ```

Should it be `refreshTransactionsSaga` or we indeed request statements? In case it has to request statements, can we reuse actions and sagas from Statements sagas and actions?
___
*[packages/cluster-ui/src/transactionsPage/transactionsPage.selectors.ts, line 12 at r1](https://reviewable.io/reviews/cockroachdb/ui/325#-MahTNoD47UZrlVPWmnn:-MahTNoD47UZrlVPWmno:b-18yf0a) ([raw file](https://github.com/cockroachdb/ui/blob/064944432cb7111aebb80304e809173bfd8c0428/packages/cluster-ui/src/transactionsPage/transactionsPage.selectors.ts#L12)):*
> ```XML
> export const selectTransactionsData = createSelector(
>   selectTransactionsSlice,
>   transactionsState => transactionsState.data || [],
> ```

this selector returns `IStatementsResponse` interface which isn't an array of objects. Defaulting to empty array is one of causes for type mismatch in Connected component.
I suggest to return `transactionsState.data` always and validate on null values later in connected components
___
*[packages/cluster-ui/src/transactionsPage/transactionsPageConnected.tsx, line 26 at r1](https://reviewable.io/reviews/cockroachdb/ui/325#-MahTBlq4hPfaez3m1Js-r1-26:-MahUYQP7Vq3lGASkqcF:bq9ictr) ([raw file](https://github.com/cockroachdb/ui/blob/064944432cb7111aebb80304e809173bfd8c0428/packages/cluster-ui/src/transactionsPage/transactionsPageConnected.tsx#L26)):*
<details><summary><i>Previously, nathanstilwell (Nathan Stilwell) wrote…</i></summary><blockquote>

@elkmaster @koorosh I'm getting a TypeScript error here, 

```
(alias) selectTransactionsData(state: AppState): any[] | cockroach.server.serverpb.StatementsResponse
import selectTransactionsData
No overload matches this call.
  The last overload gave the following error.
    Argument of type '(state: AppState) => { data: any[] | cockroach.server.serverpb.StatementsResponse; nodeRegions: { [key: string]: string; }; error: Error; }' is not assignable to parameter of type 'MapStateToPropsParam<TransactionsPageProps, RouteComponentProps<{}, StaticContext, unknown>, DefaultRootState>'.
      Type '(state: AppState) => { data: any[] | cockroach.server.serverpb.StatementsResponse; nodeRegions: { [key: string]: string; }; error: Error; }' is not assignable to type 'MapStateToPropsFactory<TransactionsPageProps, RouteComponentProps<{}, StaticContext, unknown>, DefaultRootState>'.
        Type '{ data: any[] | cockroach.server.serverpb.StatementsResponse; nodeRegions: { [key: string]: string; }; error: Error; }' is not assignable to type 'MapStateToProps<TransactionsPageProps, RouteComponentProps<{}, StaticContext, unknown>, DefaultRootState>'.
          Type '{ data: any[] | StatementsResponse; nodeRegions: { [key: string]: string; }; error: Error; }' provides no match for the signature '(state: DefaultRootState, ownProps: RouteComponentProps<{}, StaticContext, unknown>): TransactionsPageProps'.ts(2769)
index.d.ts(213, 5): The last overload is declared here.
```

Not convinced I have this right yet, but I don't see anything missing from the Transactions Page props. 
</blockquote></details>

@nathanstilwell ,  the first type for generic `connect` function should be `TransactionsPageStateProps` instead of `TransactionsPageProps`. This caused type mismatch (and see comment above for `selectTransactionsData` function.

**Review by nathanstilwell — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/ui/325)* status: all files reviewed, 3 unresolved discussions (waiting on @elkmaster, @koorosh, and @nathanstilwell)

___
*[packages/cluster-ui/src/store/transactions/transactions.sagas.ts, line 8 at r1](https://reviewable.io/reviews/cockroachdb/ui/325#-MahVqMD3XvoX2RBs2z1:-Maid72N5C7rUQqV5LFe:b-sokq4z) ([raw file](https://github.com/cockroachdb/ui/blob/064944432cb7111aebb80304e809173bfd8c0428/packages/cluster-ui/src/store/transactions/transactions.sagas.ts#L8)):*
<details><summary><i>Previously, koorosh (Andrii Vorobiov) wrote…</i></summary><blockquote>

Should it be `refreshTransactionsSaga` or we indeed request statements? In case it has to request statements, can we reuse actions and sagas from Statements sagas and actions?
</blockquote></details>

Yes, I think that should be `refreshTransactionsSaga`. I missed a rename. One thing I was initially confused about is that Transactions seem to use the same request as Statements (and the response types are the same) so I considered re-using Statements sagas, but thought it would be easier to maintain if transactions had their own.

**Review by nathanstilwell — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/ui/325)* status: all files reviewed, 3 unresolved discussions (waiting on @elkmaster and @koorosh)

___
*[packages/cluster-ui/src/transactionsPage/transactionsPage.selectors.ts, line 12 at r1](https://reviewable.io/reviews/cockroachdb/ui/325#-MahTNoD47UZrlVPWmnn:-Maj-l36EK3iy9sbkRe5:bofyz0n) ([raw file](https://github.com/cockroachdb/ui/blob/064944432cb7111aebb80304e809173bfd8c0428/packages/cluster-ui/src/transactionsPage/transactionsPage.selectors.ts#L12)):*
<details><summary><i>Previously, koorosh (Andrii Vorobiov) wrote…</i></summary><blockquote>

this selector returns `IStatementsResponse` interface which isn't an array of objects. Defaulting to empty array is one of causes for type mismatch in Connected component.
I suggest to return `transactionsState.data` always and validate on null values later in connected components

</blockquote></details>

Right. Thanks for the catch.

**Comment by nathanstilwell** [comment]
___
*[packages/cluster-ui/src/transactionsPage/transactionsPage.selectors.ts, line 12 at r1](https://reviewable.io/reviews/cockroachdb/ui/325#-MahTNoD47UZrlVPWmnn:-Maj00kjA6x3LlHIm77v:barpcdi) ([raw file](https://github.com/cockroachdb/ui/blob/064944432cb7111aebb80304e809173bfd8c0428/packages/cluster-ui/src/transactionsPage/transactionsPage.selectors.ts#L12)):*
<details><summary><i>Previously, nathanstilwell (Nathan Stilwell) wrote…</i></summary><blockquote>

Right. Thanks for the catch.
</blockquote></details>

I think I was looking at the `statementsPage.selectors` when I wrote this,

```ts
  statementsSelector,
  statementsState => {
    if (!statementsState.data) {
      return [];
    }
```

the `statementsState.data` is a `StatementsResponse` (which implements and `IStatementsResponse`) and also doesn't seem like an array. Is this a bug or am I missing something?

**Comment by nathanstilwell** [comment]
___
*[packages/cluster-ui/src/transactionsPage/transactionsPageConnected.tsx, line 26 at r1](https://reviewable.io/reviews/cockroachdb/ui/325#-MahTBlq4hPfaez3m1Js-r1-26:-Maj1xFFCe6IQTrxydYa:b-y33205) ([raw file](https://github.com/cockroachdb/ui/blob/064944432cb7111aebb80304e809173bfd8c0428/packages/cluster-ui/src/transactionsPage/transactionsPageConnected.tsx#L26)):*
<details><summary><i>Previously, koorosh (Andrii Vorobiov) wrote…</i></summary><blockquote>

@nathanstilwell ,  the first type for generic `connect` function should be `TransactionsPageStateProps` instead of `TransactionsPageProps`. This caused type mismatch (and see comment above for `selectTransactionsData` function.
</blockquote></details>

Clearly, I was just tired. :smile: I refactored those types for the explicit purpose of using `TransactionsPageStateProps`. Thanks for the extra set of eyes.

**Comment by nathanstilwell** [comment]
Cluster UI has been migrated to [cockroachdb/cockroach](https://github.com/cockroachdb/cockroach) so this PR is replaced by https://github.com/cockroachdb/cockroach/pull/66036, https://github.com/cockroachdb/cockroach/pull/66044, and https://github.com/cockroachdb/cockroach/pull/66049

**Pull request closed without merging**